File: src/main/java/ch/wisv/choice/exam/model/Exam.java
package ch.wisv.choice.exam.model;


import ch.wisv.choice.course.model.Course;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.Setter;

import javax.persistence.*;
import java.time.LocalDate;

/**
 * An Exam entity represents an exam that was taken for a course,
 * not the physical document with the questions on it.
 */
@Table(uniqueConstraints = @UniqueConstraint(columnNames = {"course", "date"}))
@Entity
@AllArgsConstructor
public class Exam {

    @Getter
    @Id
    @GeneratedValue
    private Long id;

    @Getter
    @Setter
    @ManyToOne(targetEntity = Course.class)
    @JoinColumn(name = "course")
    public Course course;

    @Getter
    @Setter
    @JoinColumn(name = "date")
    public LocalDate date;

    @Getter
    @Setter
    // Used to extinguish between different types of exams: Tentamen, Hertentamen, Deeltentamen, etc..
    // Should not include date or course code.
    public String name;
}

File: src/main/kotlin/ch/wisv/choice/exam/model/Exam.kt
/*
 * Copyright (c) 2016  W.I.S.V. 'Christiaan Huygens'
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package ch.wisv.choice.exam.model

import ch.wisv.choice.course.model.Course
import com.fasterxml.jackson.annotation.JsonFormat
import java.time.LocalDate
import javax.persistence.*

@Table(uniqueConstraints = arrayOf(UniqueConstraint(columnNames = arrayOf("course", "name", "date"))))
@Entity
data class Exam(

        @Id
        @GeneratedValue
        val id: Long? = null,

        @ManyToOne(targetEntity = Course::class)
        @JoinColumn(name = "course")
        var course: Course = Course(),

        @JoinColumn(name = "date")
        var date: LocalDate = LocalDate.now(),

        var includingAnswers: Boolean = false,

        // Used to extinguish between different types of exams: Exam, Resit, Mid term, Answers, etc..
        // Should not include date or course code.
        var name: String = ""
)
